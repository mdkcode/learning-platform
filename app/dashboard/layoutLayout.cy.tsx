import React from "react";
import Layout from "./layout";
import "@/app/globals.css";
import { AppRouterContext } from "next/dist/shared/lib/app-router-context.shared-runtime";
import { SessionProvider } from "next-auth/react";
import * as nextRouter from "next/navigation";
import { mockUserSession } from "@/app/configs/utils/mocks";

describe("<Layout />", () => {
  it("should redirect to home if the session is invalid", () => {
    const router = {
      push: cy.stub().as("router:push"),
      back: cy.stub(),
      forward: cy.stub(),
      refresh: cy.stub(),
      replace: cy.stub(),
      prefetch: cy.stub(),
    };
    // Mock useRouter to return the mocked router
    cy.stub(nextRouter, "useRouter").returns(router);

    cy.mount(
      <AppRouterContext value={router}>
        <SessionProvider session={mockUserSession}>
          <Layout>
            <div>Test Child</div>
          </Layout>
        </SessionProvider>
      </AppRouterContext>
    );
    cy.contains("Test Child").should("exist");
    cy.get('[data-test="mobileHeader"]').should("exist");
    cy.get('[data-test="sidenav"]').should("exist");
  });
});
