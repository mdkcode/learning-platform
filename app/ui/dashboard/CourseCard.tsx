"use client";
import { getCourseList } from "@/app/api/courses/courses.api";
import {
  CourseProps,
  CourseSearchParams,
} from "@/app/api/courses/courses.interface";
import {
  DehydratedState,
  HydrationBoundary,
  useQuery,
} from "@tanstack/react-query";

interface CourseCardsListClientProps extends CourseSearchParams {
  dehydratedState: DehydratedState;
}

export const CourseCardsList = ({
  searchQuery,
  dehydratedState,
}: CourseCardsListClientProps) => {
  const { data: videos } = useQuery({
    queryKey: ["courses", searchQuery],
    queryFn: () => getCourseList(searchQuery),
  });

  return (
    <HydrationBoundary state={dehydratedState}>
      <div className="flex gap-5 flex-wrap">
        {videos?.map((video: CourseProps) => (
          <CourseCard key={video.id} {...video} />
        ))}
      </div>
    </HydrationBoundary>
  );
};

export const CourseCard = ({ name, description }: CourseProps) => {
  return (
    <div
      className="w-[400px] inline-block rounded-lg p-5 bg-gradient-to-b bg-gray-200 
      relative cursor-pointer transition-all duration-300 transform 
      hover:scale-105 hover:shadow-lg hover:from-gray-200"
    >
      <h3 className="text-xl font-bold my-2">{name}</h3>
      <p className="mb-8">{description}</p>
      <div className="absolute bottom-3">
        <p className="text-lg text-blue-500">Subscribe</p>
      </div>
    </div>
  );
};
