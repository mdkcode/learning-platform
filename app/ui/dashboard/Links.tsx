import { Routes } from "@/app/routes/routes";
import Link from "next/link";
import Icon from "@/app/ui/components/Icon/Icon";

const menu = [
  {
    label: "Dashboard",
    href: Routes.DASHBOARD,
  },
  {
    label: "My courses",
    href: Routes.COURSES,
  },
];

export function Links() {
  return (
    <div className="text-white text-center">
      {menu.map(({ label, href }) => (
        <Link key={label} href={href}>
          <p className="font-medium text-lg my-4">{label}</p>
        </Link>
      ))}
      <Link href={Routes.HOME}>
        <div className="flex justify-center gap-1 items-center mb-2">
          <p className="font-medium text-lg">Exit</p>
          <Icon name="exit" className="text-white" />
        </div>
      </Link>
    </div>
  );
}
