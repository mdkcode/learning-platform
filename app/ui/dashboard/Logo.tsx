export function Logo() {
  return (
    <h1 className="text-2xl font-bold my-1 bg-gradient-to-r from-orange-400 to-pink-500 inline-block text-transparent bg-clip-text">
      SkillEd
    </h1>
  );
}
