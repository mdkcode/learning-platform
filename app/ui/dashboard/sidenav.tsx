import { Routes } from "@/app/routes/routes";
import { ArrowRightStartOnRectangleIcon } from "@heroicons/react/24/outline";
import Link from "next/link";

const menu = [
  {
    label: "Home",
    href: Routes.DASHBOARD,
  },
  {
    label: "My courses",
    href: Routes.COURSES,
  },
];

export default function SideNav() {
  return (
    <div className="hidden h-screen py-4 md:block md:px-2 w bg-[#00004B] text-white w-[10%] text-center">
      <h1 className="text-2xl font-bold my-1 bg-gradient-to-r from-orange-400 to-pink-500 inline-block text-transparent bg-clip-text">
        SkillEd
      </h1>
      {menu.map(({ label, href }) => (
        <Link key={label} href={href}>
          <p className="font-medium text-lg my-4">{label}</p>
        </Link>
      ))}
      <Link href={Routes.HOME}>
        <div className="flex justify-center gap-1 items-center">
          <p className="font-medium text-lg my-4">Exit</p>
          <ArrowRightStartOnRectangleIcon className="size-6" />
        </div>
      </Link>
    </div>
  );
}
