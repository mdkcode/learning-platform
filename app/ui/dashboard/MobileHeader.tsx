"use client";
import { useState } from "react";
import Icon from "../components/Icon/Icon";
import { Logo } from "./Logo";
import { Links } from "./Links";

export function MobileHeader() {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  return (
    <>
      <div className="bg-[#00004B] flex justify-between items-center md:hidden p-5">
        <Logo />
        <button onClick={() => setIsOpen((prev) => !prev)}>
          <Icon name={isOpen ? "close" : "hamburger"} className="text-white" />
        </button>
      </div>
      {isOpen && (
        <div className="bg-[#00004B] border-t py-4">
          <Links />
        </div>
      )}
    </>
  );
}
