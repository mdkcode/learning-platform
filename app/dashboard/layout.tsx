"use client";
import { useSession } from "next-auth/react";
import { MobileHeader } from "../ui/dashboard/MobileHeader";
import SideNav from "../ui/dashboard/sidenav";
import { useRouter } from "next/navigation";
import { useEffect } from "react";
import { Routes } from "../routes/routes";

const Layout = ({ children }: { children: React.ReactNode }) => {
  const router = useRouter();
  const { data: session } = useSession();

  useEffect(() => {
    if (session?.expires && !session.user) router.push(Routes.HOME);
  }, [session?.expires, session?.user, router]);

  return (
    <div className="flex flex-col md:h-screen md:flex-row">
      <MobileHeader />
      <SideNav />
      <div className="flex-grow p-6 md:overflow-y-auto md:p-12">{children}</div>
    </div>
  );
};

export default Layout;
