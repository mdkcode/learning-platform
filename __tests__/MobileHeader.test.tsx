import { render, fireEvent, screen } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import { MobileHeader } from "@/app/ui/dashboard/MobileHeader";

describe("MobileHeader", () => {
  it("matches snapshot when menu is closed", () => {
    const { container } = render(<MobileHeader />);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot when menu is open", () => {
    const { container } = render(<MobileHeader />);
    const [toggleButton] = screen.getAllByTestId("mobileMenuButton");
    fireEvent.click(toggleButton);
    expect(container).toMatchSnapshot();
  });
});
