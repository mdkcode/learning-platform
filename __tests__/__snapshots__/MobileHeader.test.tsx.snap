// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MobileHeader > matches snapshot when menu is closed 1`] = `
<div>
  <div
    class="bg-[#00004B] flex justify-between items-center md:hidden p-5"
    data-test="mobileHeader"
  >
    <h1
      class="text-2xl font-bold my-1 bg-gradient-to-r from-orange-400 to-pink-500 inline-block text-transparent bg-clip-text"
    >
      SkillEd
    </h1>
    <button
      data-testid="mobileMenuButton"
    >
      <svg
        aria-hidden="true"
        class="w-6 h-6 text-black-500 text-white"
        data-slot="icon"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`MobileHeader > matches snapshot when menu is open 1`] = `
<div>
  <div
    class="bg-[#00004B] flex justify-between items-center md:hidden p-5"
    data-test="mobileHeader"
  >
    <h1
      class="text-2xl font-bold my-1 bg-gradient-to-r from-orange-400 to-pink-500 inline-block text-transparent bg-clip-text"
    >
      SkillEd
    </h1>
    <button
      data-testid="mobileMenuButton"
    >
      <svg
        aria-hidden="true"
        class="w-6 h-6 text-black-500 text-white"
        data-slot="icon"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</div>
`;
